### PR TITLE
feature: report cpu info to noderesoucetopology

### LINKED ISSUE #264: [proposal] report cpu-topology and cpuset allocs for cpu fine-grained orchestration
**What is your proposal**:
report these field in NUMAResourceTopology

- node.koordinator.sh/cpu-topology
- node.koordinator.sh/cpu-shared-pools
- node.koordinator.sh/pod-cpu-allocs（without origin guaranteed-static）

**Why is this needed**:
see https://github.com/koordinator-sh/koordinator/issues/232

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: Yue Zhang <huaihuan.zy@alibaba-inc.com>

### Ⅱ. Does this pull request fix one issue?
fix #264 

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/statesinformer/config.go [hunk 1]
```diff
@@ -47,4 +52,6 @@ func (c *Config) InitFlags(fs *flag.FlagSet) {
 	fs.DurationVar(&c.KubeletSyncTimeout, "kubelet-sync-timeout", c.KubeletSyncTimeout, "The length of time to wait before giving up on a single request to Kubelet. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h).")
 	fs.BoolVar(&c.InsecureKubeletTLS, "kubelet-insecure-tls", c.InsecureKubeletTLS, "Using read-only port to communicate with Kubelet. For testing purposes only, not recommended for production use.")
 	fs.UintVar(&c.KubeletReadOnlyPort, "kubelet-read-only-port", c.KubeletReadOnlyPort, "The read-only port for the kubelet to serve on with no authentication/authorization. Default: 10255.")
+	fs.StringVar(&c.KubeletPreferredAddressType, "KubeletPreferredAddressType", c.KubeletPreferredAddressType, "The node address types to use when determining which address to use to connect to a particular node.")
+	fs.DurationVar(&c.NodeTopologySyncInterval, "NodeTopologySyncInterval", c.NodeTopologySyncInterval, "The interval which Koordlet will report the node topology info")
```
Comment (jasonliu747) [inline pkg/koordlet/statesinformer/config.go:56]: `NodeTopologySyncInterval` -> `node-topology-sync-interval`

**Review by jasonliu747 — CHANGES_REQUESTED** [review]

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		if len(cpus) <= 0 {
+			continue
+		}
+		set := cpuset.NewCPUSet(cpus...)
+		sharePools = append(sharePools, extension.CPUSharedPool{
+			CPUSet: set.String(),
+			Node:   nodeID,
+			Socket: cpuUsage[int32(cpus[0])].Socket,
+		})
+	}
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:156]: can be abstracted as a function

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+	cpuSharePoolsJson, err := json.Marshal(sharePools)
+	if err != nil {
+		klog.Errorf("failed to marshal cpushare pools of node %s, err: %v", nodeName, err)
+		return
+	}
+
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:164]: Conflicts generally do not occur, but other errors may be encountered, such as `TooManyRequests`

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		topology, err := s.topologyClient.TopologyV1alpha1().NodeResourceTopologies().Get(ctx, topologyName, metav1.GetOptions{})
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:165]: recommend set ResourceVersion with "0" in metav1.GetOptions to query the topology from APIServer cache

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+			klog.Errorf("failed to get node resource topology %s, err: %v", topologyName, err)
+			return err
+		}
+		if topology.Annotations == nil {
+			topology.Annotations = make(map[string]string)
+		}
+		topology.Annotations[extension.AnnotationNodeCPUTopology] = string(cpuTopologyJson)
+		topology.Annotations[extension.AnnotationNodeCPUAllocs] = string(podAllocsJson)
+		topology.Annotations[extension.AnnotationNodeCPUSharedPools] = string(cpuSharePoolsJson)
+		_, err = s.topologyClient.TopologyV1alpha1().NodeResourceTopologies().Update(context.TODO(), topology, metav1.UpdateOptions{})
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:176]: Before updating, we should check the difference between current and old to reduce the number of updates.

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+func (s *statesInformer) syncNodeResourceTopology(node *corev1.Node) {
+	topologyName := fmt.Sprintf("koord-%s", node.Name)
+	ctx := context.TODO()
+	blocker := true
+	_, err := s.topologyClient.TopologyV1alpha1().NodeResourceTopologies().Get(ctx, topologyName, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("failed to get node resource topology %s, err: %v", topologyName, err)
+			return
+		}
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:46]: suggetion:

```go
if err == nil {
  return
}

if !errors.IsNotFound(err) {
   return
}

// create new topology here
```

File: pkg/koordlet/statesinformer/states_noderesourcetopology_test.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	r.syncNodeResourceTopology(testNode)
+
+	topologyName := fmt.Sprintf("koord-%s", testNode.Name)
+
+	topology, err := client.TopologyV1alpha1().NodeResourceTopologies().Get(context.TODO(), topologyName, metav1.GetOptions{})
+
+	assert.Equal(t, nil, err)
+	assert.Equal(t, topologyName, topology.Name)
+	assert.Equal(t, "Koordinator", topology.Labels["app.kubernetes.io/managed-by"])
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology_test.go:55]: replace the label key with extension.LabelManagedBy

File: pkg/koordlet/statesinformer/states_noderesourcetopology_test.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+		topologyClient: client,
+		metricsCache:   mockMetricCache,
+		podMap:         mockPodMeta,
+		node:           testNode,
+	}
+
+	r.reportNodeTopology()
+
+	topology, err := client.TopologyV1alpha1().NodeResourceTopologies().Get(context.TODO(), topologyName, metav1.GetOptions{})
+	assert.Equal(t, nil, err)
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology_test.go:140]: It is clearer to define a value called `expectTopology` for comparison.

**Review by eahydra — CHANGES_REQUESTED** [review]

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+						Kind:               "Node",
+						Name:               node.Name,
+						UID:                node.GetUID(),
+						Controller:         &blocker,
+						BlockOwnerDeletion: &blocker,
+					},
+				},
+			},
+		}
+		_, err = s.topologyClient.TopologyV1alpha1().NodeResourceTopologies().Create(ctx, topology, metav1.CreateOptions{})
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:65]: the fields `topologyPolicies` and `zones` are required. This creation operation will fail. It is recommended to add an E2E test or verify it by yourself.

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+)
+
+func (s *statesInformer) syncNodeResourceTopology(node *corev1.Node) {
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:36]: add #TODO retry even Node obj is not updated

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+	_, err = s.topologyClient.TopologyV1alpha1().NodeResourceTopologies().Create(ctx, topology, metav1.CreateOptions{})
+	if err != nil {
+		klog.Errorf("failed to create node resource topology %s, err: %v", topologyName, err)
+		return
+	}
+
+}
+
+func (s *statesInformer) reportNodeTopology() {
+	nodeName := s.node.Name
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:79]: s.GetNode() with mutex

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+	}
+	_, err = s.topologyClient.TopologyV1alpha1().NodeResourceTopologies().Create(ctx, topology, metav1.CreateOptions{})
+	if err != nil {
+		klog.Errorf("failed to create node resource topology %s, err: %v", topologyName, err)
+		return
+	}
+
+}
+
+func (s *statesInformer) reportNodeTopology() {
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:78]: better to divide this function into three functions:
1. generate cpuTopo
2. calcu cpu share pool
3. update crd

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+			continue
+		}
+		if status.CPUSet != "" {
+			set, err := cpuset.Parse(status.CPUSet)
+			if err != nil {
+				klog.Errorf("failed to parse cpuset info of pod %s, err: %v", podMeta.Pod.Name, err)
+				continue
+			}
+			for _, cpuID := range set.ToSliceNoSort() {
+				usedCPUs[int32(cpuID)] = nil
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:120]: delete usedCPUs[int32(cpuID)]

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+					Kind:               "Node",
+					Name:               node.Name,
+					UID:                node.GetUID(),
+					Controller:         &blocker,
+					BlockOwnerDeletion: &blocker,
+				},
+			},
+		},
+		// fields are required
+		TopologyPolicies: []string{"None"},
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:67]: https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/blob/master/pkg/apis/topology/v1alpha1/types.go#L25 defined the const value None

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+)
+
+func (s *statesInformer) syncNodeResourceTopology(node *corev1.Node) {
+	topologyName := fmt.Sprintf("koord-%s", node.Name)
```
Comment (stormgbs) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:37]: It is better to re-using node name for NodeResourceTopology like NodeMetrics, NodeSLO, etc.

Comment (eahydra) [reply pkg/koordlet/statesinformer/states_noderesourcetopology.go:37]: In order to be compatible with the existing NodeResourceTopology instances and prevent koordlet from conflicting with existing components.

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		klog.Errorf("failed to create node resource topology %s, err: %v", topologyName, err)
+		return
+	}
+
+}
+
+func (s *statesInformer) reportNodeTopology() {
+	nodeName := s.node.Name
+	topologyName := fmt.Sprintf("koord-%s", nodeName)
```
Comment (stormgbs) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:80]: Name style.

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+					Controller:         &blocker,
+					BlockOwnerDeletion: &blocker,
+				},
+			},
+		},
+		// fields are required
+		TopologyPolicies: []string{"None"},
+		Zones:            v1alpha1.ZoneList{v1alpha1.Zone{Name: "fake-name", Type: "fake-type"}},
+	}
+	_, err = s.topologyClient.TopologyV1alpha1().NodeResourceTopologies().Create(ctx, topology, metav1.CreateOptions{})
```
Comment (stormgbs) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:70]: Call create or update based on existence of NodeResourceTopology.

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		klog.Errorf("failed to create node resource topology %s, err: %v", topologyName, err)
```
Comment (stormgbs) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:72]: node resource topology ->  NodeResourceTopology

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	podMetas := s.GetAllPods()
+	for _, podMeta := range podMetas {
+		status, err := extension.GetResourceStatus(podMeta.Pod.Annotations)
+		if err != nil {
+			klog.Errorf("failed to get resource status of pod %s, err: %v", podMeta.Pod.Name, err)
+			continue
+		}
+		if status.CPUSet != "" {
```
Comment (stormgbs) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:113]: Fast-Return may be better:  if status.CPUSet == "" { continue }

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+					Name:               node.Name,
+					UID:                node.GetUID(),
+					Controller:         &blocker,
+					BlockOwnerDeletion: &blocker,
+				},
+			},
+		},
+		// fields are required
+		TopologyPolicies: []string{"None"},
+		Zones:            v1alpha1.ZoneList{v1alpha1.Zone{Name: "fake-name", Type: "fake-type"}},
```
Comment (stormgbs) [inline pkg/koordlet/statesinformer/states_noderesourcetopology.go:68]: Does a useless NodeTopology cause some unexpected effects on existing pods, especially for LSE/LSR pods ?

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go (hunk 15 above)
Comment (hormes) [reply pkg/koordlet/statesinformer/states_noderesourcetopology.go:37]: In fact, we expect koordlet to be an alternative to topology gathering, and if that's the case, it's better to keep the name the same.

File: apis/extension/node.go [hunk 21]
```diff
@@ -77,3 +77,7 @@ func GetPodCPUAllocs(annotations map[string]string) (PodCPUAllocs, error) {
 	}
 	return allocs, nil
 }
+
+func (p PodCPUAllocs) Len() int           { return len(p) }
+func (p PodCPUAllocs) Less(i, j int) bool { return p[i].Name < p[j].Name }
+func (p PodCPUAllocs) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
```
Comment (jasonliu747) [inline apis/extension/node.go:83]: I would suggest that you move theses three functions after #L53.

File: pkg/koordlet/statesinformer/config.go [hunk 22]
```diff
@@ -25,10 +25,13 @@ import (
...
-	KubeletSyncTimeout          time.Duration
-	InsecureKubeletTLS          bool
-	KubeletReadOnlyPort         uint
+
+	KubeletSyncInterval time.Duration
+	KubeletSyncTimeout  time.Duration
+	InsecureKubeletTLS  bool
+	KubeletReadOnlyPort uint
+
+	NodeTopologySyncInterval time.Duration
```
Comment (jasonliu747) [inline pkg/koordlet/statesinformer/config.go:34]: keep it tight, new line is not needed.

File: pkg/koordlet/statesinformer/states_noderesourcetopology.go (hunk 15 above)
Comment (eahydra) [reply pkg/koordlet/statesinformer/states_noderesourcetopology.go:37]: > In fact, we expect koordlet to be an alternative to topology gathering, and if that's the case, it's better to keep the name the same.

OK。 LGTM.

**Pull request merged**